security/pfSense-pkg-nmap/files/usr/local/www/nmap_scan.php

```diff
@@ -0,0 +1,468 @@
+	if (is_array($config['openvpn']["openvpn-{$mode}"])) {
+		foreach ($config['openvpn']["openvpn-{$mode}"] as $id => $setting) {
+			if (!isset($setting['disable'])) {
+				$interfaces['ovpn' . substr($mode, 0, 1) . $setting['vpnid']] = $mode_descr . ": ".htmlspecialchars($setting['description']);
+			}
+		}
+	}
+}
+
+$interfaces = array_merge($interfaces, interface_ipsec_vti_list_all());
+$interfaces = array_merge(array('' => 'Any'), $interfaces);
+
+$scan_types = array(
+	'syn' => gettext('SYN'),
+	'connect' => gettext('TCP connect()'),
+	'icmp' => gettext('Ping'),
+	'udp' => gettext('UDP'),
+	'arp' => gettext('ARP (directly connected networks only!)')
+);
+
+function nmap_get_running_process(){
+	$processcheck = (trim(shell_exec("/bin/ps axw -O pid= | /usr/bin/grep '/usr/local/bin/nmap' | /usr/bin/grep '{$fn}' | /usr/bin/egrep -v '(pflog|grep)'")));
+	return $processcheck;
+}
+
+// only accept alphanumeric, spaces (040) and '.,-_' chars, and do not allow '-o? ' for output files
+function valid_input($str) {
+	$valid_str = (preg_match('/^[a-z0-9\040\.,\-_]+$/i', $str) === 1) && (preg_match('/-o.\040/', $str) === 0);
+	//$valid_str = (preg_match('/^[a-z0-9\040\.,\-_]+$/i', $str) === 1) && (strpos($str, '-o') === false);
+	return $valid_str;
+}
+
+$input_errors = array();
+$do_nmapscan = false;
+if ($_POST) {
+	$hostname = $_POST['hostname'];
+	$interface = $_POST['interface'];
+	$scantype = $_POST['scantype'];
+	$noping = isset($_POST['noping']);
+	$servicever = isset($_POST['servicever']);
+	$osdetect = isset($_POST['osdetect']);
+	$custom_scantype = $_POST['custom_scantype'];
+	$custom_options = $_POST['custom_options'];
+	$custom_targetspecs = $_POST['custom_targetspecs'];
+
+	if ($_POST['startbtn'] != "") {
+		$action = gettext("Start");
+
+		// check for input errors
+		if (empty($hostname)) {
+			$input_errors[] = gettext("You must enter an IP address to scan.");
+		} elseif (!(is_ipaddr($hostname) || is_subnet($hostname) || is_hostname($hostname))) {
+			$input_errors[] = gettext("You must enter a valid IP address to scan.");
+		}
+
+		if(!empty($interface)) {
+			if (!array_key_exists($interface, $interfaces)) {
+				$input_errors[] = gettext("Invalid interface.");
+			}
+		}
+
+		// process scan options
+		if (!count($input_errors)) {
+
+			// prevent use of any deprecated option
+			$nmap_options = " -d";
+
+			if (is_ipaddrv6($hostname) || is_subnetv6($hostname)) {
+				$nmap_options .= " -6";
+			}
+
+			// scan type
+			if (empty($custom_scantype)) {
+				switch($scantype) {
+					case 'syn':
+						$nmap_options .= " -sS";
+						break;
+					case 'connect':
+						$nmap_options .= " -sT";
+						break;
+					case 'icmp':
+						$nmap_options .= " -sn"; // previously -sP
+						break;
+					case 'udp':
+						$nmap_options .= " -sU";
+						break;
+					case 'arp':
+						$nmap_options .= " -sn -PR";
+						break;
+				}
+			} elseif (valid_input($custom_scantype)) {
+					$nmap_options .= " " . $custom_scantype;
```

Comment:
I know you're severely limiting the input allowed here but it still seems like a bad idea to allow arbitrary parameters to be passed.

If someone knows that much about nmap parameters they aren't likely to be using the GUI for this. Possible, but unlikely.